controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -578,28 +578,50 @@ func firstNodeNotMatchingName(name string, nodes []corev1.Node) *corev1.Node {
 func (w *Workload) UpdateKubeProxyImageInfo(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane) error {
 	ds := &appsv1.DaemonSet{}
 
-	if err := w.Client.Get(ctx, ctrlclient.ObjectKey{Name: kubeProxyDaemonSetName, Namespace: metav1.NamespaceSystem}, ds); err != nil {
+	if err := w.Client.Get(ctx, ctrlclient.ObjectKey{Name: kubeProxyKey, Namespace: metav1.NamespaceSystem}, ds); err != nil {
 		if apierrors.IsNotFound(err) {
 			// if kube-proxy is missing, return without errors
 			return nil
 		}
-		return errors.Wrapf(err, "failed to determine if %s daemonset already exists", kubeProxyDaemonSetName)
+		return errors.Wrapf(err, "failed to determine if %s daemonset already exists", kubeProxyKey)
 	}
 
-	if len(ds.Spec.Template.Spec.Containers) == 0 {
+	container := findKubeProxyContainer(ds)
+	if container == nil {
 		return nil
 	}
-	newImageName, err := util.ModifyImageTag(ds.Spec.Template.Spec.Containers[0].Image, kcp.Spec.Version)
+
+	newImageName, err := util.ModifyImageTag(container.Image, kcp.Spec.Version)
 	if err != nil {
 		return err
 	}
 
-	if ds.Spec.Template.Spec.Containers[0].Image != newImageName {
+	if container.Image != newImageName {
 		patch := client.MergeFrom(ds.DeepCopy())
-		ds.Spec.Template.Spec.Containers[0].Image = newImageName
+		patchKubeProxyImage(ds, newImageName)
 		if err := w.Client.Patch(ctx, ds, patch); err != nil {
 			return errors.Wrap(err, "error patching kube-proxy DaemonSet")
 		}
 	}
 	return nil
 }
+
+func findKubeProxyContainer(ds *appsv1.DaemonSet) *corev1.Container {
+	containers := ds.Spec.Template.Spec.Containers
+	for _, c := range containers {
+		c := c
```

Comment:
I'd rather do `for i := range containers` and `c := containers[i]` than shadow redeclare `c` this way.

Reply:
Agree. I think it's a bit more readable. Also could we do this within the if condition? Seems like an unnecessary operation `if c.Name != kubeProxyKey`